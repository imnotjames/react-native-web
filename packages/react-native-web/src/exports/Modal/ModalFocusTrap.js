/**
 * Copyright (c) Nicolas Gallagher.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import React, { useRef, useCallback, useEffect } from 'react';

import { canUseDOM } from 'fbjs/lib/ExecutionEnvironment';

import View from '../View';
import createElement from '../createElement';
import StyleSheet from '../StyleSheet';

/**
 * This Component is used to "wrap" the modal we're opening
 * so that changing focus via tab will never leave the document.
 *
 * This allows us to properly trap the focus within a modal
 * even if the modal is at the start or end of a document.
 */

const FocusBracket = () => {
  return createElement(
    'div',
    {
      style: styles.focusBracket,
      accessible: true,

      // Sets `aria-hidden` to true
      importantForAccessibility: 'no-hide-descendants',
      accessibilityRole: 'none',

      // `importantForAccessibility` being set to `no-hide-descendants`  will prevent
      // these two attributes from being set as needed.
      ['data-focusable']: true,
      tabIndex: 0
    }
  );
};

function attemptFocus(element: any) {
  try {
    element.focus();
  } catch (e) {
    // Do nothing
  }

  return document.activeElement === element;
}

function focusFirstDescendant(element: any) {
  for (let i = 0; i < element.childNodes.length; i++) {
    const child = element.childNodes[i];
    if (attemptFocus(child) || focusFirstDescendant(child)) {
      return true;
    }
  }
  return false;
}

function focusLastDescendant(element: any) {
  for (let i = element.childNodes.length - 1; i >= 0; i--) {
    const child = element.childNodes[i];
    if (attemptFocus(child) || focusLastDescendant(child)) {
      return true;
    }
  }
  return false;
}

export type ModalFocusTrapProps = {|
  children?: any,

  active?: boolean | () => boolean,
|};

const ModalFocusTrap = ({ active, children }: ModalFocusTrapProps) => {
  const trapElementRef = useRef();

  // Ref used to track trapping of focus and to prevent focus from leaving a modal
  // for accessibility reasons per W3CAG.
  const focusRef = useRef<{ trapFocusInProgress: boolean, lastFocusedElement: ?HTMLElement }>({
    trapFocusInProgress: false,
    lastFocusedElement: null
  });

  const trapFocus = useCallback(() => {
    // We should not trap focus if:
    // - The modal hasn't fully initialized with an HTMLElement ref
    // - Focus is already in the process of being trapped (eg, we're refocusing)
    // - isTrapActive prop being false-ish tells us to do nothing
    if (!trapElementRef.current || focusRef.current.trapFocusInProgress || !active) {
      return;
    }

    try {
      focusRef.current.trapFocusInProgress = true;

      // Only muck with the focus if the event target isn't within this modal
      if (document.activeElement instanceof Node && !trapElementRef.current.contains(document.activeElement)) {
        // To handle keyboard focusing we can make an assumption here.
        // If you're tabbing through the focusable elements, the previously
        // active element will either be the first or the last.
        //
        // If the previously selected element is the "first" descendant
        // and we're leaving it - this means that we should
        // be looping around to the other side of the modal.
        let hasFocused = focusFirstDescendant(trapElementRef.current);
        if (focusRef.current.lastFocusedElement === document.activeElement) {
          hasFocused = focusLastDescendant(trapElementRef.current);
        }

        // If we couldn't focus a new element then we need to blur the active element
        if (!hasFocused && document.activeElement) {
          document.activeElement.blur();
        }
      }
    } finally {
      focusRef.current.trapFocusInProgress = false;
    }

    focusRef.current.lastFocusedElement = document.activeElement;
  }, [active]);

  // Bind to the document itself for this component
  useEffect(() => {
    if (canUseDOM) {
      document.addEventListener('focus', trapFocus, true);
    }

    // Call the trapFocus callback at least once when this modal has been
    // re-rendered / trapFocus has changed!
    trapFocus();

    return () => {
      if (canUseDOM) {
        document.removeEventListener('focus', trapFocus, true);
      }
    };
  }, [trapFocus]);

  return (
    <>
      <FocusBracket />
      <View ref={trapElementRef}>
        {children}
      </View>
      <FocusBracket />
    </>
  );
};

export default ModalFocusTrap;

const styles = StyleSheet.create({
  focusBracket: {
    outlineStyle: 'none'
  }
});
