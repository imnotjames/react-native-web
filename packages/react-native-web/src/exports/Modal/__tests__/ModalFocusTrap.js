/* eslint-env jasmine, jest */

import ModalFocusTrap from '../ModalFocusTrap';
import React from 'react';
import { cleanup, render } from '@testing-library/react';

describe('components/Modal/ModalFocusTrap', () => {
  afterEach(() => cleanup());

  test('creates focusable "brackets"', () => {
    const { container } = render(
      <ModalFocusTrap>
        <a href={'#hello'}>Hello</a>
      </ModalFocusTrap>
    );

    expect(container.children.length).toBe(3);

    const beforeBracket = container.firstChild;
    const afterBracket = container.lastChild;

    expect(beforeBracket.children.length).toBe(0);
    expect(afterBracket.children.length).toBe(0);

    expect(beforeBracket.getAttribute('aria-hidden')).toBe('true');
    expect(beforeBracket.getAttribute('tabindex')).toBe('0');

    expect(afterBracket.getAttribute('aria-hidden')).toBe('true');
    expect(afterBracket.getAttribute('tabindex')).toBe('0');

    expect(container).toMatchSnapshot();
  });

  test('focus is not trapped by default', () => {
    const { getByTestId } = render(
      <>
        <a data-testid={'outside'} href={'#outside'}>Outside</a>
        <ModalFocusTrap>
          <a data-testid={'inside'} href={'#hello'}>Hello</a>
        </ModalFocusTrap>
      </>
    );

    const outsideElement = getByTestId('outside');

    outsideElement.focus();

    expect(document.activeElement).toBe(outsideElement);
  });

  test('focus is trapped when active flag changes', () => {
    const { getByTestId, rerender } = render(
      <>
        <a data-testid={'outside'} href={'#outside'}>Outside</a>
        <ModalFocusTrap active={false}>
          <a data-testid={'inside'} href={'#hello'}>Hello</a>
        </ModalFocusTrap>
      </>
    );

    getByTestId('outside').focus();

    expect(document.activeElement).toBe(getByTestId('outside'));

    rerender(
      <>
        <a data-testid={'outside'} href={'#outside'}>Outside</a>
        <ModalFocusTrap active={true}>
          <a data-testid={'inside'} href={'#hello'}>Hello</a>
        </ModalFocusTrap>
      </>
    );

    expect(document.activeElement).toBe(getByTestId('inside'));
  });

  test('focus is trapped when active', () => {
    const { getByTestId } = render(
      <>
        <a data-testid={'outside'} href={'#outside'}>Outside</a>
        <ModalFocusTrap active={true}>
          <a data-testid={'inside'} href={'#hello'}>Hello</a>
        </ModalFocusTrap>
      </>
    );

    const outsideElement = getByTestId('outside');
    const insideElement = getByTestId('inside');

    outsideElement.focus();

    expect(document.activeElement).toBe(insideElement);
  });

  test('focus is trapped without contents', () => {
    const { getByTestId } = render(
      <>
        <a data-testid={'outside'} href={'#outside'}>Outside</a>
        <ModalFocusTrap active>
          <div data-testid={'inside'}>There are no focusable contents.</div>
        </ModalFocusTrap>
      </>
    );

    const outsideElement = getByTestId('outside');

    outsideElement.focus();

    expect(document.activeElement).toBe(document.body);
  });

  test('focus is not trapped when inactive', () => {
    const { getByTestId } = render(
      <>
        <a data-testid={'outside'} href={'#outside'}>Outside</a>
        <ModalFocusTrap active={false}>
          <a data-testid={'inside'} href={'#hello'}>Hello</a>
        </ModalFocusTrap>
      </>
    );

    const outsideElement = getByTestId('outside');

    outsideElement.focus();

    expect(document.activeElement).toBe(outsideElement);
  });
});
