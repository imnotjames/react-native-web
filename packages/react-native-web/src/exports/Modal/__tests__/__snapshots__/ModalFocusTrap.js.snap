// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Modal/ModalFocusTrap creates focusable "brackets" 1`] = `
<div>
  <div
    aria-hidden="true"
    class="r-outlineStyle-1ny4l3l"
    data-focusable="true"
    role="presentation"
    tabindex="0"
  />
  <div
    class="css-view-1dbjc4n"
  >
    <a
      href="#hello"
    >
      Hello
    </a>
  </div>
  <div
    aria-hidden="true"
    class="r-outlineStyle-1ny4l3l"
    data-focusable="true"
    role="presentation"
    tabindex="0"
  />
</div>
`;
